Client: Refactor client

SimonRichardson opened the pull request:
The following change firstly bumps juju/http to v2, as this contains a
lot of breaking changes.

Secondly it introduces the concept of transport middleware. In order to
add additional middlewares in the future (thinking about retry and
metric transport wrappers), then we need to expose this to the consumer
of the library.

Part of that standardisation was to correctly setup a default transport
so that all subsquent transports use the same one. Instead of some using
a blank transport and others using a custom one.

The middleware concept then allows us to talk about concepts like
breakers, that prevent outgoing calls at a much higher level.

Lastly we moved from a fat config, to providing options that can be
sequenced togther, before passing it to the constructor. One benefit of
doing it this way is allowing us to create defaults in one location and
pass that through, rather than ad-hoc inline style.

SimonRichardson committed:
Client: Refactor client

The following change firstly bumps juju/http to v2, as this contains a
lot of breaking changes.

Secondly it introduces the concept of transport middleware. In order to
add additional middlewares in the future (thinking about retry and
metric transport wrappers), then we need to expose this to the consumer
of the library.

Part of that standardisation was to correctly setup a default transport
so that all subsquent transports use the same one. Instead of some using
a blank transport and others using a custom one.

The middleware concept then allows us to talk about concepts like
breakers, that prevent outgoing calls at a much higher level.

Lastly we moved from a fat config, to providing options that can be
sequenced togther, before passing it to the constructor. One benefit of
doing it this way is allowing us to create defaults in one location and
pass that through, rather than ad-hoc inline style.

hmlanigan commented on client.go:
```diff
@@ -21,14 +22,16 @@ import (
+// touch!
 func init() {
 	defaultTransport := http.DefaultTransport.(*http.Transport)
 	// Call the HTTPDialShim for the DefaultTransport to
```
HTTPDialShim - comment needs updating

hmlanigan commented on client.go:
```diff
@@ -42,99 +45,181 @@ type Logger interface {
+	}
+}
+
+// WithLogger defines a logger to use with the client.
```
Add a note that it's recommended to use a child of your current logger.

hmlanigan commented on tls.go:
```diff
@@ -9,23 +9,53 @@ import (
+func DefaultHTTPTransportWithMiddlewares(middlewares []TransportMiddleware) *http.Transport {
+	return NewHTTPTLSTransport(TransportConfig{
+		DisableKeepAlives:   true,
+		TLSHandshakeTimeout: 30 * time.Second,
```
You're increasing the timeout value here from the original.  Should we all the HandshakeTimeout, and others to be configurable?  Perhaps allow it as a model config value eventually?

SimonRichardson replied:
OK yeah, agreed!

hmlanigan reviewed (approved):
A few comments, overall looks good.  A great update to how we wrap the http package.

SimonRichardson committed:
Transport: Set keep-alives and handshake timeout

It seems reasonable to use the options made to actually set on the http
transport.

SimonRichardson commented:
$$merge$$
